### DISCUSSION TITLE
Does a component know its constrained size before Render()?

### DISCUSSION THREAD

**Discussion by jsholmes** [ArthurSonzogni/FTXUI#423] [Q&A] [answered]
I have a view into a very large datagrid that I'm writing and it would be nice if it knew how large of a viewport it has to use.  What's the best way to get this to happen?  The main container is giving it a size() constraint like `size(HEIGHT, EQUAL, rows)`.

But I don't see how the component knows that constraint when it renders.

**Comment by ArthurSonzogni** [comment, marked as answer]
Hello!

A component is responsible for creating the DOM (aka element) via Render(). Then the element is responsible for drawing on the screen 

Hence the following modules of FTXUI.
- component/
- element/
- screen/

So a component can't know the size of a element it created, before drawing it on the screen 

However, you can know the size of an element after it has been drawn, using the `reflect` decorator.
https://arthursonzogni.github.io/FTXUI/namespaceftxui.html#aaff8245861617a3d9e846e99de582a63

```cpp
Box box;
auto wrapped = element | reflect(&box);
```

Note that box will be populated after the `wrapped` has been rendered on the screen.
So you will be able to use the information of the previous frame to build the next one. That's what most components uses to check for mouse.

**Reply by ArthurSonzogni** [reply]
If your question was about knowing the terminal dimension, you can use `Terminal::Size()`.

**Reply by jsholmes** [reply]
Awesome thanks.  reflect(&box) is great.
